File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;
import java.io.IOException;

public class Main {
	static int N, M, start;
	static List<Integer>[] link;
	static boolean[] visited;

	public static void main(String[] args) throws IOException {
		// 입력
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());

		start = Integer.parseInt(st.nextToken()) - 1;
		link = new ArrayList[N];
		for (int i = 0; i < N; i++) {
			link[i] = new ArrayList<>(); // 각 정점의 인접 리스트 초기화
		}

		for (int i = 0; i < M; i++) {
			st = new StringTokenizer(br.readLine());
			int n1 = Integer.parseInt(st.nextToken()) - 1;
			int n2 = Integer.parseInt(st.nextToken()) - 1;
			link[n1].add(n2);
			link[n2].add(n1);
		}
		////// 입력완료

		for (int i = 0; i < N; i++) {
			Collections.sort(link[i]);
		}

		visited = new boolean[N];
		dfs(start);
		System.out.println();
		visited = new boolean[N];
		bfs(start);

	}

	private static void bfs(int start) {
		Queue<Integer> queue = new ArrayDeque<>();
		queue.add(start);
		visited[start] = true;

		do {
			int n = queue.poll();
			System.out.print(n + 1 + " ");
			for (int i = 0; i < link[n].size(); i++) {
				if (!visited[link[n].get(i)]) {
					visited[link[n].get(i)] = true;
					queue.add(link[n].get(i));
				}
			}
		} while (!queue.isEmpty());

	}

	private static void dfs(int n) {
		visited[n] = true;
		System.out.print(n + 1 + " ");
		for (int num : link[n]) {
			if (!visited[num]) {
				dfs(num);
			}
		}

	}

}

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
# 시작시간 : 9시 18분
# 끝난 시간 : 9시 27분
# 소요시간 : 약 9분

# 정점 번호가 작은 것을 먼저 방문

def dfs(cur):
    visited[cur] = 1
    dfs_ans.append(cur)
    for nxt in adj[cur]:
        if visited[nxt]:  # 이미 방문 했으면
            continue
        dfs(nxt)
        visited[nxt] = 1


V, E, S = map(int, input().split())
adj = [[] for _ in range(V + 1)]

for _ in range(E):
    a, b = map(int, input().split())
    adj[a].append(b)
    adj[b].append(a)  # 양방향

# 정렬
for i in range(1, V + 1):
    adj[i].sort()

# dfs
dfs_ans = []
visited = [0] * (V + 1)
dfs(S)

# bfs
bfs_ans = []
visited = [0] * (V + 1)

from collections import deque

queue = deque([S])
visited[S] = 1
bfs_ans.append(S)

while queue:
    q = queue.pop()
    for nxt in adj[q]:
        if visited[nxt]:  # 이미 방문 했으면
            continue
        queue.appendleft(nxt)
        bfs_ans.append(nxt)
        visited[nxt] = 1

print(*dfs_ans)
print(*bfs_ans)
